services/libs/integrations/src/integrations/youtube/processStream.ts

```diff
@@ -0,0 +1,74 @@
+// processStream.ts content
+import { YoutubeRootStream, YoutubeIntegrationSettings, YoutubeVideoStreamConfig } from './types'
+import { getVideos } from './api/videos'
+import { getComments } from './api/comments'
+import { getVideosByKeywords } from './api/videosByKeywords'
+import { IProcessStreamContext, ProcessStreamHandler } from '../../types'
+
+const handler: ProcessStreamHandler = async (ctx) => {
+  if (ctx.stream.identifier.startsWith(YoutubeRootStream.UPLOADED_VIDEOS)) {
+    await handleChannelStream(ctx)
+  } else if (ctx.stream.identifier.startsWith(YoutubeRootStream.KEYWORDS_SEARCH)) {
+    await handleKeywordsSearchStream(ctx)
+  } else {
+    await handleVideoStream(ctx)
+  }
+}
+
+async function handleChannelStream(ctx: IProcessStreamContext) {
+  const channelSettings = ctx.stream.data as YoutubeIntegrationSettings
+  const videos = await getVideos(ctx)
+
+  for (const video of videos.items) {
+    const videoId = video.id.videoId
+
+    await ctx.publishStream(`${YoutubeRootStream.CHANNEL_VIDEO}:${videoId}`, {
+      title: video.snippet.title,
+      videoId,
+      apiKey: channelSettings.apiKey
+    })
+  }
+
+  const shouldLoadNextPage = videos.nextPageToken && videos.nextPageToken != ''
+  if (shouldLoadNextPage) {
+    await ctx.publishStream(
+      `${YoutubeRootStream.UPLOADED_VIDEOS}:${channelSettings.channelId}:${channelSettings.nextPageToken}`,
+      {
+        ...channelSettings,
+        nextPageToken: videos.nextPageToken
+      },
+    )
+  }
+}
+
+async function handleKeywordsSearchStream(ctx: IProcessStreamContext) {
```

Comment:
Keywords search still returns a huge number of results should we relax it a bit by putting a cap on the number of pages to fetch ?

Reply:
Yeah that would make sense. Do you have an idea what the number should be?

Reply:
@garrrikkotua I was thinking maybe we can let the user have more control by allowing them to add more than one api key if they need to, and let the number of pages be defaulted to 40, that would cost 4k credits for both channel based and search api, if the user wants to increase the pages number they can do it from the UI. 

The api keys settings attributes representation would be something like: 

```json
{
"apiKeys": [{
  "value": String,
  "resetAt": ISODate
}]
}
```

Once the api returns an error we can pick the api key based on the date.

This would be ideal in case you want the user to have more control over the feature but will need more implementation code. Let me know what do you think.